src/NServiceBus.Testing.Fakes/Fakes/FakeBuilder.cs

```diff
@@ -0,0 +1,152 @@
+        /// </summary>
+        public void Register<T>(params T[] instance) where T: class
+        {
+            instances.Add(typeof(T), instance);
+        }
+
+        /// <summary>
+        /// Registers a factory method which will be invoked every time an instance of <typeparamref name="T"/> is resolved.
+        /// </summary>
+        public void Register<T>(Func<T> factory)
+        {
+            factories.Add(typeof(T), () => new object[]
+            {
+                factory()
+            });
+        }
+
+        /// <summary>
+        /// Registers a factory method which will be invoked every time an instance of <typeparamref name="T"/> is resolved.
+        /// </summary>
+        public void Register<T>(Func<T[]> factory) where T: class
+        {
+            factories.Add(typeof(T), factory);
+        }
+
+        /// <summary>
+        /// Returns an instantiation of the given type.
+        /// </summary>
+        /// <param name="typeToBuild">The <see cref="T:System.Type"/> to build.</param>
+        /// <returns>
+        /// The component instance.
+        /// </returns>
+        public virtual object Build(Type typeToBuild)
+        {
+            return BuildAll(typeToBuild).First();
+        }
+
+        /// <summary>
+        /// Creates an instance of the given type, injecting it with all defined dependencies.
+        /// </summary>
+        /// <typeparam name="T">Type to be resolved.</typeparam>
+        /// <returns>
+        /// Instance of <typeparamref name="T"/>.
+        /// </returns>
+        public virtual T Build<T>()
+        {
+            return BuildAll<T>().First();
+        }
+
+        /// <summary>
+        /// For each type that is compatible with the given type, an instance is created with all dependencies injected.
+        /// </summary>
+        /// <param name="typeToBuild">The <see cref="T:System.Type"/> to build.</param>
+        /// <returns>
+        /// The component instances.
+        /// </returns>
+        public virtual IEnumerable<object> BuildAll(Type typeToBuild)
+        {
+            if (instances.ContainsKey(typeToBuild))
+            {
+                return instances[typeToBuild];
+            }
+
+            if (factories.ContainsKey(typeToBuild))
+            {
+                return factories[typeToBuild]();
+            }
+
+            throw new Exception($"Cannot build instance of type {typeToBuild} because there was no instance of factory registered for it.");
+        }
+
+        /// <summary>
+        /// For each type that is compatible with T, an instance is created with all dependencies injected, and yielded to the
+        /// caller.
+        /// </summary>
+        /// <typeparam name="T">Type to be resolved.</typeparam>
+        /// <returns>
+        /// Instances of <typeparamref name="T"/>.
+        /// </returns>
+        public virtual IEnumerable<T> BuildAll<T>()
```

Comment:
@andreasohlund exactly, we think these are good defaults, but still want users to have freedom to do their own thing.